### ISSUE TITLE
Can’t get it to run in a chess GUI 

### ISSUE THREAD

**Issue by tosca07** [mattbruv/Turochamp#2]
Hello Matt,

As it is a WIP project I don’t know if the engine should already work. I have create a windows binary but can’t get to work in a chess GUI which accepts UCI engines like ARENA, SHREDDER Or Fritz.

Is it not yet finished or should it already work in chess GUIs like the ones mentioned above?

Thanks & regards
Dirk

**Comment by mattbruv** [comment]
Hey Dirk,

I haven't worked on this project in quite some time. It's unfinished. I definitely want to get around to finishing it, because it is a valuable piece of chess history that hasn't been made into an open source program. I stopped working on it because I had a hard time interpreting how certain details of the evaluation and search functions should be written. 

I can't make any promises on when or if I will finish it because it is not one of my top priorities at the moment. I will say that I do eventually plan on making progress on it at some point. Sorry for the vague answer, but that's the only answer I can give you right now.

**Comment by tosca07** [comment]
Hi Matt,

Thanks for your quick answer!

Yes it is indeed a valuable piece of history and an open source uci engine would be really cool and a lot of people were searching such an engine  already and the only other one one done by ChessBase for Fritz Long Time ago is no longer working in the current Fritz and ChessBase GUI.

What a pity that’s hasn’t been finished but if course I understand that priorities can shift...

By the way I found a Turochamp uci implementation here on github in python:
https://github.com/mdoege/PyTuroChamp
That one works but it is not user friendly as you have to install python and the engine is because of that way too slow - sometimes it takes more than 5 minutes for one move - almost so time consuming like the real paper machine 😏

Maybe it is an idea to have a look at that python script regarding the evaluation functions (i m not a programmer unfortunately) when you decide to work again on this project .

I will have a look at this project from time to time because I find your work really interesting and valuable...

Thanks again for answering - I appreciate that!

Regards from Germany
Dirk

**Comment by mattbruv** [comment]
Thanks for sharing that link. It looks like that project was started just a few months ago, so that's why I didn't know about it. However in his readme, he says:

> Pyturochamp.py does not actually reproduce the results of either the Turing paper or the Chessbase implementation for Fritz. But then again Turing's paper was meant as a proof-of-concept, so maybe reproducibility is not important.

It doesn't look like his goal was to accurately reproduce the Turing engine (or the Chessbase version), but rather to use the evaluation features of the Turing engine as a starting point for his own simple engine. I will have a look through his code though and see if I can take anything away from it.

The more I think about it, it might be a good idea to get the engine working as much as I can, even if it is slightly incorrect. Then I can spread the word and advertise the engine. Maybe at that point some people who are interested could contribute and help iron out any incorrect details. That is probably the smart way to go about it.

Even the programmers at chessbase [had a hard time](https://en.chessbase.com/post/reconstructing-turing-s-paper-machine) figuring out how to interpret it apparently: 

> The ChessBase team spent some weeks re-examining the Turing papers and discussing the implementation in the Turing engine. Since they could not locate the error they consulted Ken Thompson, one of the pioneers of computer chess (and author of Unix and C). He spent some time wrestling with the problem, and in fact wrote his own code on the basis of his reading of the Turing instructions. His program produced most of the ChessBase deviations from the Glennie game. It was quite baffling.

So it sounds like the implementations of the engine even varied between chessbase and Ken Thompson.

I think at some point I will sit down and try to get a very basic engine with as much as I can implement that makes sense to me. Once I do that, It will be much easier for others to want to contribute I think.

**Comment by tosca07** [comment]
Yep I think even a basic Turochamp UCI engine would be very welcome and coagulation lead to a more and more accurate version.

At the moment the engine does not return any output from any chess gui i own  - it just does not nothing... :-(

Looks like each implementation approach differs someway and most of the implementations offer some enhancements you could switch on to make a more playable game for andvanced users but at least they offer all one mode which should be very near to original Turing heuristic.

I could Verfahren that with python Turochamp and the iOS Turochamp app (which is free) - they make almost the same moves (only the sequence differs from time to time).

I would be happy to support testing...

So long - all the best, Matt.

Thanks again for your work!
Dirk

**Comment by mdoege** [comment]
If I may chime in here, yes, I think it is difficult to program an engine that will exactly reproduce the moves from Turing's sample game. Because (a) Turing did not calculate all his moves correctly, starting with 1. e4 and (b) some things in Turing's algorithm are open to interpretation and probably quite slow if implemented (e.g. his dead position criterion). That is why I put that sort of disclaimer in my project description.

That said, I think pyturochamp.py should be reasonably close to the moves from Turing's program, especially if you allow for some tuning with the piece-square table. I know speed could be better (I think with PyPy it's not too bad), so if Matt's Rust version would work that would be very nice of course.

The PyTuroChamp chess engines are mostly meant for easy fiddling with their code, but execution speed is not really their forte.

**Comment by tosca07** [comment]
Hi Martin,
cool that you chimed in. So I can say a big thank you to you as well because PyTuroChamp is the only working UCI engine I could use for my chess guis.
On my notebook it really runs very slow (more than 3-5 minutes when more moves are possible (I set PSTAB=0 to get as close to the original turochamp). I downloaded peppy but could not get it work correctly as I am not an expert with all this (I just want to use it but could support testing etc.)

Thats why a new compiled UCI version would be really great (I also own the chessbase/Fritz turochamp engine but they won't work in newer ochessbase or Fritz GUIs because of a new changed engine format).
Maybe I try to get peppy working (at the moment it does not grin python chess with pypy3)...

Thanks you both for your time spending one this interesting "historical" chess engine.

Dirk

**Comment by mdoege** [comment]
What is the problem with python-chess and pypy you are experiencing? Maybe I can help if you describe your issue in more detail. Perhaps you should open a new issue in my repo though as I did not intend to take over this thread in Matt's repo for discussing my project. Also, we can talk in German there if you prefer...
